elements/lisk-p2p/test/integration/p2p.ts

```diff
@@ -1349,47 +1308,186 @@ describe('Integration tests for P2P library', () => {
-
 					const peerPorts = [...newPeers, ...triedPeers].map(
 						peerInfo => peerInfo.wsPort,
 					);
 
 					expect(ALL_NODE_PORTS_WITH_LIMIT).to.include.members(peerPorts);
+				}
+			});
+		});
+
+		describe('P2P.request', () => {
+			beforeEach(() => {
+				for (let p2p of p2pNodeList) {
+					// Collect port numbers to check which peer handled which request.
+					p2p.on('requestReceived', request => {
+						if (!request.wasResponseSent) {
+							request.end({
+								nodePort: p2p.nodeInfo.wsPort,
+								requestProcedure: request.procedure,
+								requestData: request.data,
+								requestPeerId: request.peerId,
+							});
+						}
+					});
+				}
+			});
+
+			it('should make request to the network; it should reach a single peer', async () => {
+				const firstP2PNode = p2pNodeList[0];
+				const response = await firstP2PNode.request({
+					procedure: 'foo',
+					data: 'bar',
 				});
+				expect(response).to.have.property('data');
+				expect(response.data)
+					.to.have.property('nodePort')
+					.which.is.a('number');
+				expect(response.data)
+					.to.have.property('requestProcedure')
+					.which.is.a('string');
+				expect(response.data)
+					.to.have.property('requestData')
+					.which.is.equal('bar');
+				expect(response.data)
+					.to.have.property('requestPeerId')
+					.which.is.equal(`127.0.0.1:${firstP2PNode.nodeInfo.wsPort}`);
+			});
+
+			// Check for even distribution of requests across the network. Account for an error margin.
+			it('requests made to the network should be distributed randomly', async () => {
+				const TOTAL_REQUESTS = 1000;
+				const firstP2PNode = p2pNodeList[0];
+				const nodePortToResponsesMap: any = {};
+
+				const expectedAverageRequestsPerNode =
+					TOTAL_REQUESTS / NETWORK_PEER_COUNT_WITH_LIMIT;
+				const expectedRequestsLowerBound = expectedAverageRequestsPerNode * 0.4;
+				const expectedRequestsUpperBound = expectedAverageRequestsPerNode * 1.6;
+
+				for (let i = 0; i < TOTAL_REQUESTS; i++) {
+					const response = await firstP2PNode.request({
+						procedure: 'foo',
+						data: i,
+					});
+					let resultData = response.data as any;
+					if (!nodePortToResponsesMap[resultData.nodePort]) {
+						nodePortToResponsesMap[resultData.nodePort] = [];
+					}
+					nodePortToResponsesMap[resultData.nodePort].push(resultData);
+				}
+
+				for (let requestsHandled of Object.values(
+					nodePortToResponsesMap,
+				) as any) {
+					expect(requestsHandled).to.be.an('array');
+					expect(requestsHandled.length).to.be.greaterThan(
+						expectedRequestsLowerBound,
+					);
+					expect(requestsHandled.length).to.be.lessThan(
+						expectedRequestsUpperBound,
+					);
+				}
+			});
+		});
+
+		describe('P2P.send', () => {
+			let collectedMessages: Array<any> = [];
+
+			beforeEach(() => {
+				collectedMessages = [];
+				for (let p2p of p2pNodeList) {
+					p2p.on('messageReceived', message => {
+						collectedMessages.push({
```

Comment:
Here we're just recording the messages. It would be good to add another test case for propagation.

We could add a special propagation handler like this:

```js
let propagationStart = Date.now();
p2p.on('messageReceived', message => {
  let propagationDuration = Date.now() - propagationStart;
  // All nodes will propagate messages to each other during 200ms
  if (message.event === 'propagate' && propagationDuration < 200) {
     p2p.send({ event: 'propagate', data: message.data + 1 });
  }
});
```

Then inside the test case you can do (just once; this simulates forging a block and then sending it to the network):

```
firstP2PNode.send({ event: 'propagate', data: 0 });
```

Reply:
Also, we should simulate an additional case related to what happens if a node receives a message which it has already seen (this is what happens if our node has already seen a particular block); in that case we should not propagate the message again.

To do this, we should keep a Map or Set of all messages received during the test case. The first time we see a message, we propagate it, otherwise we do nothing.